File: iosClient/WoG/WoG/ThirdwebAuth.swift
//
//  ThirdwebAuth.swift
//  WoG
//
//  Native email OTP auth via thirdweb HTTP API,
//  then exchanges the thirdweb token for a shard JWT.
//

import Foundation

struct AuthResult {
    let walletAddress: String
    let shardToken: String
    let isNewUser: Bool
}

enum ThirdwebAuth {
    static let clientId = "231a06443d1568f83d2d4f2c8e7dfe3b"
    private static let thirdwebBase = "https://api.thirdweb.com/v1/auth"
    private static let shardBase = "https://worldofgeneva.com"
    // Spoof origin so thirdweb ties initiate+complete to the same allowed domain
    private static let origin = "https://worldofgeneva.com"

    private static func thirdwebHeaders() -> [(String, String)] {
        [
            ("x-client-id", clientId),
            ("Content-Type", "application/json"),
            ("Origin", origin),
            ("Referer", origin + "/"),
            ("x-bundle-id", Bundle.main.bundleIdentifier ?? "com.worldofgeneva.WoG"),
        ]
    }

    /// Step 1: Send OTP to email
    static func sendEmailOTP(email: String) async throws {
        var request = URLRequest(url: URL(string: "\(thirdwebBase)/initiate")!)
        request.httpMethod = "POST"
        for (key, val) in thirdwebHeaders() { request.setValue(val, forHTTPHeaderField: key) }

        let body: [String: Any] = ["method": "email", "email": email]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0

        if status >= 400 {
            let msg = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw AuthError.apiError("Failed to send code (\(status)): \(msg)")
        }
    }

    /// Step 2: Verify OTP → thirdweb token + wallet
    /// Step 3: Exchange thirdweb token for shard JWT
    static func verifyEmailOTP(email: String, code: String) async throws -> AuthResult {
        var twRequest = URLRequest(url: URL(string: "\(thirdwebBase)/complete")!)
        twRequest.httpMethod = "POST"
        for (key, val) in thirdwebHeaders() { twRequest.setValue(val, forHTTPHeaderField: key) }

        let twBody: [String: Any] = ["method": "email", "email": email, "code": code]
        twRequest.httpBody = try JSONSerialization.data(withJSONObject: twBody)

        let (twData, twResponse) = try await URLSession.shared.data(for: twRequest)
        let twStatus = (twResponse as? HTTPURLResponse)?.statusCode ?? 0

        if twStatus >= 400 {
            let msg = String(data: twData, encoding: .utf8) ?? "Unknown error"
            throw AuthError.apiError("Verification failed (\(twStatus)): \(msg)")
        }

        guard let twJson = try JSONSerialization.jsonObject(with: twData) as? [String: Any],
              let walletAddress = twJson["walletAddress"] as? String,
              let thirdwebToken = twJson["token"] as? String else {
            throw AuthError.apiError("Invalid response from thirdweb")
        }

        let isNewUser = twJson["isNewUser"] as? Bool ?? false

        let shardToken = try await exchangeForShardToken(thirdwebToken: thirdwebToken)

        return AuthResult(walletAddress: walletAddress, shardToken: shardToken, isNewUser: isNewUser)
    }

    private static func exchangeForShardToken(thirdwebToken: String) async throws -> String {
        var request = URLRequest(url: URL(string: "\(shardBase)/auth/verify-thirdweb")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["thirdwebToken": thirdwebToken]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0

        if status >= 400 {
            let msg = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw AuthError.apiError("Shard auth failed (\(status)): \(msg)")
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["token"] as? String else {
            throw AuthError.apiError("Invalid response from shard auth")
        }

        return token
    }

    enum AuthError: LocalizedError {
        case apiError(String)
        var errorDescription: String? {
            switch self {
            case .apiError(let msg): return msg
            }
        }
    }
}

File: androidClient/app/src/main/java/com/worldofgeneva/app/auth/ThirdwebAuth.kt
package com.worldofgeneva.app.auth

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

data class AuthResult(
    val walletAddress: String,
    val shardToken: String,
    val isNewUser: Boolean,
)

class ThirdwebAuthException(message: String) : IOException(message)

object ThirdwebAuth {
    const val CLIENT_ID = "231a06443d1568f83d2d4f2c8e7dfe3b"
    private const val THIRDWEB_BASE = "https://api.thirdweb.com/v1/auth"
    private const val SHARD_BASE = "https://worldofgeneva.com"
    private const val ORIGIN = "https://worldofgeneva.com"
    private const val BUNDLE_ID = "com.worldofgeneva.app"

    private val JSON = "application/json".toMediaType()
    private val json = Json { ignoreUnknownKeys = true }

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    @Serializable
    private data class ThirdwebCompleteResponse(
        val walletAddress: String? = null,
        val token: String? = null,
        val isNewUser: Boolean? = null,
    )

    @Serializable
    private data class ShardVerifyResponse(val token: String? = null)

    private fun Request.Builder.thirdwebHeaders(): Request.Builder = this
        .header("x-client-id", CLIENT_ID)
        .header("Content-Type", "application/json")
        .header("Origin", ORIGIN)
        .header("Referer", "$ORIGIN/")
        .header("x-bundle-id", BUNDLE_ID)

    suspend fun sendEmailOTP(email: String) = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("method", "email")
            put("email", email)
        }.toString().toRequestBody(JSON)

        val request = Request.Builder()
            .url("$THIRDWEB_BASE/initiate")
            .post(body)
            .thirdwebHeaders()
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                val msg = response.body?.string() ?: "Unknown error"
                throw ThirdwebAuthException("Failed to send code (${response.code}): $msg")
            }
        }
    }

    suspend fun verifyEmailOTP(email: String, code: String): AuthResult = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("method", "email")
            put("email", email)
            put("code", code)
        }.toString().toRequestBody(JSON)

        val request = Request.Builder()
            .url("$THIRDWEB_BASE/complete")
            .post(body)
            .thirdwebHeaders()
            .build()

        val parsed = client.newCall(request).execute().use { response ->
            val payload = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ThirdwebAuthException("Verification failed (${response.code}): $payload")
            }
            json.decodeFromString(ThirdwebCompleteResponse.serializer(), payload)
        }

        val wallet = parsed.walletAddress
            ?: throw ThirdwebAuthException("Invalid response from thirdweb (no walletAddress)")
        val thirdwebToken = parsed.token
            ?: throw ThirdwebAuthException("Invalid response from thirdweb (no token)")

        val shardToken = exchangeForShardToken(thirdwebToken)
        AuthResult(walletAddress = wallet, shardToken = shardToken, isNewUser = parsed.isNewUser == true)
    }

    private suspend fun exchangeForShardToken(thirdwebToken: String): String = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("thirdwebToken", thirdwebToken)
        }.toString().toRequestBody(JSON)

        val request = Request.Builder()
            .url("$SHARD_BASE/auth/verify-thirdweb")
            .post(body)
            .header("Content-Type", "application/json")
            .build()

        client.newCall(request).execute().use { response ->
            val payload = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ThirdwebAuthException("Shard auth failed (${response.code}): $payload")
            }
            val parsed = json.decodeFromString(ShardVerifyResponse.serializer(), payload)
            parsed.token ?: throw ThirdwebAuthException("Invalid response from shard auth")
        }
    }
}
